CMakeLists.txt

```diff
@@ -0,0 +1,83 @@
+cmake_minimum_required(VERSION 3.0)
+
+project(sol2)
+
+set(CMAKE_CXX_STANDARD 14)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+include_directories(.)
+
+set(CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}/cmake/Modules" "${CMAKE_MODULE_PATH}")
+
+set(LUA_VERSION "5.3" CACHE STRING "The version of Lua needed (5.1, 5.2, 5.3, luaJIT)")
+if (LUA_VERSION MATCHES "5.1")
+  find_package(Lua 5.1 EXACT REQUIRED)
+elseif(LUA_VERSION MATCHES "5.2")
+  find_package(Lua 5.2 EXACT REQUIRED)
+elseif(LUA_VERSION MATCHES "5.3")
+  find_package(Lua 5.3 EXACT REQUIRED)
+elseif(LUA_VERSION MATCHES "LuaJIT")
+  find_package(LuaJIT REQUIRED)
+else()
+  message(FATAL_ERROR "${LUA_VERSION} is not a supported version for lua.")
+endif()
+
+include_directories(${LUA_INCLUDE_DIR})
+
+# Features a C++ compiler must have to be used to compile sol2
+# This list is not *complete* as CMake does not support features for
+# all of the advanced features utilized.
+set(CXX_FEATURES  cxx_auto_type
+                  cxx_constexpr
+                  cxx_decltype
+                  cxx_decltype_auto
+                  cxx_default_function_template_args
+                  cxx_final
+                  cxx_lambdas
+                  cxx_noexcept
+                  cxx_nullptr
+                  cxx_override
+                  cxx_range_for
+                  cxx_return_type_deduction
+                  cxx_right_angle_brackets
+                  cxx_static_assert
+                  cxx_strong_enums
+                  cxx_variadic_macros
+                  cxx_variadic_templates)
+
+file(GLOB HEADER_SRCS sol/*.hpp)
+set(HEADER_SRCS sol.hpp ${HEADER_SRCS})
+
+if (NOT MSVC)
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-unknown-warning -Wno-unknown-warning-option -Wall -Wextra -Wpedantic -pedantic -pedantic-errors -Wno-noexcept-type -ftemplate-depth=1024")
```

Comment:
We should have Release/Debug optimization flags. See my previous comment.

Reply:
The default target are not good enough?

`cmake . -DCMAKE_BUILD_TYPE=Release|Debug`

Reply:
CMake won't do anything on its own with the CMAKE_BUILD_TYPE. You still have to add the optimization flags yourself in the CMakeLists.txt. If you scroll up in the comments there is an example how it could look like.

Reply:
https://gitlab.kitware.com/cmake/cmake/blob/master/Modules/Compiler/GNU.cmake#L39
It does for me.

Reply:
Looks like I'm wrong, proceed :D